text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,604 @@
+a list of all properties found in the CloudFormation Resource Specification at the time of that release, and branching code-generation based on
+whether the property is in the list or not.
+
+In order to gradually enahnce the developer experience in languages without _type unions_, a new `awslint` rule will be introduced to prohibit
+introduction of new inline _type unions_. Exceptions will be added for all existing APIs.
+
+## Generating _union-like_ types
+
+Code generation needs to be changed in `jsii-pacmak` so that languages without native support for _type unions_ (**C#**, **Go**, and **Java**) have
+additional _union-like_ types emitted. The proposed API for a union such as `export type UnionType = Foo | Map<Bar> | Baz[]` in each currently
+supported language without native _type unions_ is the following (subject to change):
+
+- In **C#**
+
+  ```csharp
+  // UnionType.cs
+  public sealed class UnionType {
+    public static UnionType FromFoo(Foo value) { /* ... */ }
+    public static UnionType FromMapOfBar(IReadOnlyDictionary<Bar> value) { /* ... */ }
+    public static UnionType FromListOfBaz(IReadOnlyList<Baz> value) { /* ... */ }
+
+    // Those may throw InvalidCastException
+    public Foo Foo { get; }
+    public IReadOnlyDictionary<Bar> MapOfBar { get; }
+    public IReadOnlyList<Baz> ListOfBaz { get; }
+
+    // Private constructor ensures all initialization happens though the factories
+    private UnionType(/*...*/) { /*...*/ }
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public abstract class UnionType {
+    public static UnionType fromFoo(Foo value) {
+      return new UnionType.FooUnionType(value);
+    }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) {
+      return new UnionType.MapOfBarUnionType(value);
+    }
+    public static UnionType fromListOfBaz(Baz... value) {
+      return fromListOfBaz(Arrays.asList(value));
+    }
+    public static UnionType fromListOfBaz(List<Baz> value) {
+      return new UnionType.ListOfBaz(value);
+    }
+
+    public Foo asFoo() { throw new ClassCastException(); }
+    public Map<String, Bar> asMapOfBar() { throw new ClassCastException(); }
+    public List<Baz> asListOfBaz() { throw new ClassCastException(); }
+
+    // Private constructor ensures all initialization happens through the factories
+    private UnionType(/* ... */) { /* ... */ }
+
+    private static final class FooUnionType extends UnionType { /*...*/ }
+    private static final class MapOfBarUnionType extends UnionType { /*...*/ }
+    private static final class ListOfBazUnionType extends UnionType { /*...*/ }
+  }
+  ```
+
+The _jsii Kernel_ will be aware of the _aliased type unions_, and values passed from **JavaScript** to the _host application_ will be annotated with
+the _union type_'s fully qualified name if the value is not an instance of a known class (this is the case for values of _struct_ types). On the other
```

Comment:
> the _union type_'s fully qualified name if the value is [an object literal]

I don't think this will work, and break the wire protocol unnecessarily.

Looks like currently, the union types are a proposal that mostly affects statically typed host languages without native type union support. In the type:

```ts
type SomeUnion = string | Foo | IBar;
```

* A `string` would be serialized as `{ "jsii$native": "string", "value": "some_string" }` (or whatever the serialization is), same as today.
* A `Foo` would be serialized as `{ "jsii$objid": "Foo@123" }`, same as today

For both those cases, the only thing a language kernel would need to do is deserialize as usual, and then wrap the value in a `SomeUnion` instance.

If we were to switch object literals from:

```
{ "jsii$objid": "Object@123", "interfaces": ["IBar"] }

// to

{ "jsii$objid": "SomeUnion@123" }
```

This doesn't actually help the language kernel deserialize the right value. Worse, it now has broken the Python kernel, which used to be able to deserialize the value fine but now needs to grow support for union types (or know to ignore them, or something).

Union types should have zero effect on the wire protocol, only on the kernels on the respective sides.

Reply:
The point is to continue serializing definitely typed stuff in the same way they currently are, and to only use the union's FQN in the object ID when that is not possible (i.e: when the value returned is a struct or anonymous interface implementation).